rhizome/lib/vm_setup.rb

```diff
@@ -156,27 +158,52 @@ def parse_routes(routes)
   end
 
   def routes4(ip4, ip4_local)
-    return unless ip4
-    vm_sub = NetAddr::IPv4Net.parse(ip4)
+    vm_sub = NetAddr::IPv4Net.parse(ip4) if ip4
     local_ip = NetAddr::IPv4Net.parse(ip4_local)
-    vm, vetho, vethi = [vm_sub.to_s,
-      local_ip.network.to_s,
+    vm = vm_sub.to_s if ip4
+    vetho, vethi = [local_ip.network.to_s,
       local_ip.next_sib.network.to_s]
 
-    r "ip addr add #{vetho}/32 dev vetho#{q_vm}"
-    r "ip route add #{vm} dev vetho#{q_vm}"
+    vp.write_public_ipv4(vm) if ip4
+
+    r "ip addr replace #{vetho}/32 dev vetho#{q_vm}"
+    r "ip route replace #{vm} dev vetho#{q_vm}" if ip4
     r "echo 1 > /proc/sys/net/ipv4/conf/vetho#{q_vm}/proxy_arp"
 
-    r "ip -n #{q_vm} addr add #{vethi}/32 dev vethi#{q_vm}"
-    r "ip -n #{q_vm} addr add #{vm} dev tap#{q_vm}"
-    r "ip -n #{q_vm} route add #{vetho} dev vethi#{q_vm}"
-    r "ip -n #{q_vm} route add default via #{vetho} dev vethi#{q_vm}"
+    r "ip -n #{q_vm} addr replace #{vethi}/32 dev vethi#{q_vm}"
+    # default?
+    r "ip -n #{q_vm} route replace #{vetho} dev vethi#{q_vm}"
+    r "ip -n #{q_vm} route replace #{vm} dev tap#{q_vm}" if ip4
+    r "ip -n #{q_vm} route replace default via #{vetho} dev vethi#{q_vm}"
+
     r "ip netns exec #{q_vm} bash -c 'echo 1 > /proc/sys/net/ipv4/conf/vethi#{q_vm}/proxy_arp'"
     r "ip netns exec #{q_vm} bash -c 'echo 1 > /proc/sys/net/ipv4/conf/tap#{q_vm}/proxy_arp'"
+
+    r "ip -n #{q_vm} addr replace 192.168.0.1/24 dev tap#{q_vm}"
   end
 
-  def cloudinit(unix_user, public_key, private_subnets, ip4)
-    vm_sub = NetAddr::IPv4Net.parse(ip4) if ip4
+  def nat4(ip4, private_ipv4)
+    return unless ip4
+    public_sub = NetAddr::IPv4Net.parse(ip4)
+    public_ipv4 = public_sub.network.to_s
+
+    private_sub = NetAddr::IPv4Net.parse(private_ipv4)
+    private_ipv4 = private_sub.network.to_s
+
+    vp.write_nftables_conf(<<NFTABLES_CONF)
+table ip raw {
+  chain prerouting {
+    type filter hook prerouting priority raw; policy accept;
+    ip daddr #{public_ipv4} ip daddr set #{private_ipv4} notrack
+    ip saddr #{private_ipv4} ip daddr != 192.168.0.0/16 ip saddr set #{public_ipv4} notrack
```

Comment:
So, this line only makes sense when the private network in question is the 192.168.0.0/16 one....there's also 172.16.0.0/12 and 10.0.0.0/8. I think there's a set notation in nftables that shoud make it a cinch to add those.